### PR TITLE
fix some defects of setting default Message headers in MessagingMessageListenerAdapter

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
This PR aims to fix some minor defects in MessagingMessageListenerAdapter, including:

- improved the `checkHeaders()` method
- added the logic to apply `checkHeaders()` to Message element in Collection type return of method annotated with `@SendTo`
- added a unit testing case covering default header setting for `Iterable<Message<?>>` return type

### CODE REVIEW

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 1]
```diff
@@ -514,11 +515,11 @@ private Message<?> checkHeaders(Object result, String topic, @Nullable Object so
 			if (needsTopic) {
 				builder.setHeader(KafkaHeaders.TOPIC, topic);
 			}
-			if (needsCorrelation && sourceIsMessage) {
```
Comment (NathanQingyangXu) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:517]: seems the second condtion has been implied in the first condition

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 2]
```diff
@@ -514,11 +515,11 @@ private Message<?> checkHeaders(Object result, String topic, @Nullable Object so
...
+			if (needsCorrelation) {
 				builder.setHeader(this.correlationHeaderName,
 						((Message<?>) source).getHeaders().get(this.correlationHeaderName));
 			}
-			if (sourceIsMessage && reply.getHeaders().get(KafkaHeaders.REPLY_PARTITION) == null) {
```
Comment (NathanQingyangXu) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:521]: confusing. should `REPLY_PARTITION` be relevant on `source` side? On `reply` side, seems only `PARTITION` header matters.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 3]
```diff
@@ -501,24 +502,24 @@ else if (result instanceof Message) {
...
+	private Message<?> checkHeaders(Message<?> reply, @Nullable String topic, @Nullable Object source) { // NOSONAR (complexity)
 		MessageHeaders headers = reply.getHeaders();
-		boolean needsTopic = headers.get(KafkaHeaders.TOPIC) == null;
+		boolean needsTopic = topic != null && headers.get(KafkaHeaders.TOPIC) == null;
 		boolean sourceIsMessage = source instanceof Message;
-		boolean needsCorrelation = headers.get(this.correlationHeaderName) == null && sourceIsMessage;
-		boolean needsPartition = headers.get(KafkaHeaders.PARTITION) == null && sourceIsMessage
+		boolean needsCorrelation = sourceIsMessage && headers.get(this.correlationHeaderName) == null
+				&& ((Message<?>) source).getHeaders().get(this.correlationHeaderName) != null;
+		boolean needsPartition = sourceIsMessage && headers.get(KafkaHeaders.PARTITION) == null
```
Comment (NathanQingyangXu) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:511]: I moved `xx != null` testing to the beginning for it is much more light-weight than `headers.get(...) != null`, so shot-circuiting might avoid some unnecessary cpu cycles. Someone even claims such nullness checking involves almost zero cost in openJDK.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java [hunk 4]
```diff
@@ -501,24 +502,24 @@ else if (result instanceof Message) {
 		}
 	}
 
-	private Message<?> checkHeaders(Object result, String topic, @Nullable Object source) { // NOSONAR (complexity)
-		Message<?> reply = (Message<?>) result;
+	private Message<?> checkHeaders(Message<?> reply, @Nullable String topic, @Nullable Object source) { // NOSONAR (complexity)
```
Comment (NathanQingyangXu) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:505]: I think choosing `Message<?>` as the first argument type makes more sense, for this method is only meant to be used with that precondition.

Comment (NathanQingyangXu) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/adapter/MessagingMessageListenerAdapter.java:505]: not sure whether we can use `Nullable` to `topic`, but from its invocation context, seems we could never rule out the nullness possibility?

**Comment by sobychacko** [comment]
@NathanQingyangXu Thanks for the PR! It is now merged upstream.

**Comment by NathanQingyangXu** [comment]
> @NathanQingyangXu Thanks for the PR! It is now merged upstream.

thanks. I really appreciate that the PR got merged finally.

**Pull request closed without merging**